File: 433-minimum-genetic-mutation/minimum-genetic-mutation.cpp
class Solution {
public:
    int minMutation(string startGene, string endGene, vector<string> &bank) {
        queue<pair<string, int>> q;
        vector<bool> vis(bank.size());
        q.push({startGene, 0});
        while (!q.empty()) {
            string currGene = q.front().first;
            int mutations = q.front().second;
            q.pop();
            if (currGene == endGene)
                return mutations;
            for (int i = 0; i < bank.size(); ++i) {
                if (!vis[i] && onlyOneDifference(currGene, bank[i])) {
                    q.push({bank[i], mutations + 1}), vis[i] = true;
                }
            }
        }
        return -1;
    }

    bool onlyOneDifference(string &s1, string &s2) {
        int cnt = 0;
        for (int i = 0; i < 8; ++i) {
            if (s1[i] != s2[i]) {
                ++cnt;
                if (cnt == 2)
                    break;
            }
        }
        return cnt == 1;
    }
};

File: 433-minimum-genetic-mutation/minimum-genetic-mutation.java
class Solution {
    public int minMutation(String startGene, String endGene, String[] bank) {
        Queue<Pair> queue = new ArrayDeque<>();
        Map<String, Boolean> freq = new HashMap<>();
        queue.add(new Pair(startGene, 0));
        freq.put(startGene, true);
        int ans = -1;
        while (!queue.isEmpty()) {
            Pair front = queue.poll();
            if (front.gene.equals(endGene)) {
                ans = front.steps;
                break;
            }
            for (String gene : bank) {
                if (valid(front.gene, gene) && !freq.containsKey(gene)) {
                    freq.put(gene, true);
                    queue.add(new Pair(gene, front.steps + 1));
                }
            }
        }
        return ans;
    }

    boolean valid(String gene1, String gene2) {
        int cnt = 0;
        for (int i = 0; i < gene1.length(); ++i)
            if (gene1.charAt(i) != gene2.charAt(i))
                ++cnt;
        return cnt == 1;
    }
}

class Pair {
    String gene;
    int steps;

    public Pair(String gene, int steps) {
        this.gene = gene;
        this.steps = steps;
    }
}